src/daft-schema/src/dtype.rs

```diff
@@ -233,14 +238,31 @@ impl DataType {
             Self::List(field) => Ok(ArrowType::LargeList(Box::new(
                 arrow2::datatypes::Field::new("item", field.to_arrow()?, true),
             ))),
-            Self::Map(field) => Ok(ArrowType::Map(
-                Box::new(arrow2::datatypes::Field::new(
-                    "item",
-                    field.to_arrow()?,
-                    true,
-                )),
-                false,
-            )),
+            Self::Map { key, value } => {
+                let struct_type = ArrowType::Struct(vec![
+                    // We never allow null keys in maps for several reasons:
+                    // 1. Null typically represents the absence of a value, which doesn't make sense for a key.
+                    // 2. Null comparisons can be problematic (similar to how f64::NAN != f64::NAN).
+                    // 3. It maintains consistency with common map implementations in arrow (no null keys).
+                    // 4. It simplifies map operations
+                    //
+                    // This decision aligns with the thoughts of team members like Jay and Sammy, who argue that:
+                    // - Nulls in keys could lead to unintuitive behavior
+                    // - If users need to count or group by null values, they can use other constructs like
+                    //   group_by operations on non-map types, which offer more explicit control.
+                    //
+                    // By disallowing null keys, we encourage more robust data modeling practices and
+                    // provide a clearer semantic meaning for map types in our system.
+                    arrow2::datatypes::Field::new("key", key.to_arrow()?, true),
```

Comment:
I might be missing something, but why do we have the preamble that keys cannot be nullable, but then we set `is_nullable` to `true`?

Reply:
it used to be `false` .. I think there are issues with types though which is why I changed it to true.. might be fixed though with how physical types now work.

Reply:
dd4a338c4674bed53bb5ecf77915e6d0e40a96e0
 resovled